src/plugins/push/pushactivation.ts

```diff
@@ -0,0 +1,605 @@
+    return NotActivated;
+  }
+  return null;
+};
+
+const WaitingForNewPushDeviceDetails = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    machine.callActivatedCallback(null);
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof CalledDeactivate) {
+    machine.deregister();
+    return WaitingForDeregistration(WaitingForNewPushDeviceDetails);
+  } else if (event instanceof GotPushDeviceDetails) {
+    machine.updateRegistration();
+    return WaitingForRegistrationSync;
+  }
+};
+
+const WaitingForRegistrationSync = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    machine.callActivatedCallback(null);
+    return WaitingForRegistrationSync;
+  } else if (event instanceof RegistrationUpdated) {
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof SyncRegistrationFailed) {
+    machine.callUpdateRegistrationFailedCallback(event.reason);
+    return AfterRegistrationUpdateFailed;
+  }
+  return null;
+};
+
+const AfterRegistrationUpdateFailed = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate || event instanceof GotPushDeviceDetails) {
+    machine.updateRegistration();
+    return WaitingForRegistrationSync;
+  } else if (event instanceof CalledDeactivate) {
+    machine.deregister();
+    return WaitingForDeregistration(AfterRegistrationUpdateFailed);
+  }
+  return null;
+};
+
+type ActivationState = (machine: ActivationStateMachine, event: ActivationEvent) => ActivationState | null;
+
+type ActivationStateName =
+  | 'NotActivated'
+  | 'WaitingForPushDeviceDetails'
+  | 'WaitingForDeviceRegistration'
+  | 'WaitingForNewPushDeviceDetails'
+  | 'WaitingForRegistrationSync'
+  | 'AfterRegistrationUpdateFailed';
+
+export const ActivationStates = {
+  NotActivated,
+  WaitingForPushDeviceDetails,
+  WaitingForDeviceRegistration,
+  WaitingForNewPushDeviceDetails,
+  WaitingForRegistrationSync,
+  AfterRegistrationUpdateFailed,
+};
+
+const WaitingForDeregistration = function (previousState: ActivationState): ActivationState {
+  return function (machine, event) {
+    if (event instanceof CalledDeactivate) {
+      return WaitingForDeregistration(previousState);
+    } else if (event instanceof Deregistered) {
+      const device = machine.rest.device as LocalDevice;
+      delete device.deviceIdentityToken;
+      delete device.push.recipient;
+      device.resetId();
+      device.persist();
+      machine.callDeactivatedCallback(null);
+      return NotActivated;
+    } else if (event instanceof DeregistrationFailed) {
+      machine.callDeactivatedCallback(event.reason);
+      return previousState;
+    }
+    return null;
+  };
+};
+
+function isPersistentState(state: ActivationState) {
+  return state.name == 'NotActivated' || state.name == 'WaitingForNewPushDeviceDetails';
```

Comment:
I think this is gonna be a problem in the minified code, `.name` won't match expected function names (added `minified: true` to esbuild config for push plugin and all state names were gone).
can either convert state functions to actual classes (might be time consuming), or just add a custom property with hardcoded name to those state functions and compare using that.
can then change calls like `this.current.name` and `maybeNext.name` to use that new prop too.

`event.constructor.name` calls will suffer the similar fate

Reply:
i figured converting them to actual classes was the better approach, done that now